### PR TITLE
yosys-systemverilog: fix multirange with dot usage

### PR DESCRIPTION
**Pull request by kamilrakoczy**
Ref: https://github.com/chipsalliance/yosys-f4pga-plugins/issues/524

This PR adds support for multirange in ``dot`` context (before first ``dot``).

UHDM-integration-test: https://github.com/chipsalliance/UHDM-integration-tests/pull/740
yosys-systemverilog run: https://github.com/antmicro/yosys-systemverilog/actions/runs/5186512253/jobs/9347862380

### CODE REVIEW

File: systemverilog-plugin/UhdmAst.cc [hunk 1]
```diff
@@ -962,29 +962,47 @@ static AST::AstNode *convert_dot(AST::AstNode *wire_node, AST::AstNode *node, AS
...
+                      : (*it)->children[0]->clone();
+        // calculate which struct we selected
+        auto move_offset = new AST::AstNode(AST::AST_MUL, AST::AstNode::mkconst_int(struct_size_int, true, 32), elem);
+        // move our expanded dot to currently selected struct
+        expanded->children[0] = new AST::AstNode(AST::AST_ADD, move_offset->clone(), expanded->children[0]);
+        expanded->children[1] = new AST::AstNode(AST::AST_ADD, move_offset, expanded->children[1]);
+        // next range, multiply struct_size_int by whole dimension size
+        struct_size_int *= *dimensions_it;
+        // move iterator to next dimension width
+        dimensions_it += 2;
```
Comment (wsipak) [inline systemverilog-plugin/UhdmAst.cc:1003]: Shouldn't we check whether `dimensions_it` exceeds the length of `multirage_dimensions`?
We could have something to make sure that we don't violate the size of the vector.

File: systemverilog-plugin/UhdmAst.cc [hunk 2]
```diff
@@ -962,29 +962,47 @@ static AST::AstNode *convert_dot(AST::AstNode *wire_node, AST::AstNode *node, AS
...
-              new AST::AstNode(AST::AST_MUL, AST::AstNode::mkconst_int(struct_size_int, true, 32), node->children[0]->children[0]->clone()));
+    std::vector<AST::AstNode *> struct_ranges;
+    for (auto c : node->children) {
+        if (c->type == AST::AST_RANGE) {
+            struct_ranges.push_back(c);
         }
     }
+    const auto wire_node_unpacked_ranges_size = wire_node->attributes[UhdmAst::unpacked_ranges()]->children.size();
+    // 'dot' notation can only be used on specific struct element
+    // '1' packed range is from wiretype and it is already expanded in 'expand_dot'
```
Comment (mglb) [inline systemverilog-plugin/UhdmAst.cc:973]: What `'1'` means? Is it the first (leftmost) packed range?

**Review by mglb — COMMENTED** [review]
After the first read it looks OK, but I would like to look at it one more time in a few hours.

File: systemverilog-plugin/UhdmAst.cc [hunk 3]
```diff
@@ -962,29 +962,47 @@ static AST::AstNode *convert_dot(AST::AstNode *wire_node, AST::AstNode *node, AS
...
+        log_assert((*it)->children.size() == 1);
+        bool is_unpacked_range = range_id < wire_node_unpacked_ranges_size;
+        // if unpacked range, select from back;
+        auto elem = is_unpacked_range
+                      ? new AST::AstNode(AST::AST_SUB, AST::AstNode::mkconst_int(*dimensions_it - 1, true, 32), (*it)->children[0]->clone())
+                      : (*it)->children[0]->clone();
+        // calculate which struct we selected
+        auto move_offset = new AST::AstNode(AST::AST_MUL, AST::AstNode::mkconst_int(struct_size_int, true, 32), elem);
+        // move our expanded dot to currently selected struct
+        expanded->children[0] = new AST::AstNode(AST::AST_ADD, move_offset->clone(), expanded->children[0]);
```
[the remark this answers is not in the document]
Comment (kamilrakoczy) [reply systemverilog-plugin/UhdmAst.cc:998]: Yes, ``expanded`` is result from ``expand_dot`` that always return ``AST_RANGE`` with 2 children:
https://github.com/chipsalliance/yosys-f4pga-plugins/blob/8edc027cdc484f86c5ac8698c696c74d04e6f92e/systemverilog-plugin/UhdmAst.cc#L955

[the remark this answers is not in the document]
Comment (wsipak) [reply systemverilog-plugin/UhdmAst.cc:998]: Makes sense, thanks!

File: systemverilog-plugin/UhdmAst.cc (hunk 2 above)
Comment (kamilrakoczy) [reply systemverilog-plugin/UhdmAst.cc:973]: I tried to explain here why there is ``wire_node_ranges_size - 1`` in assert (it was due to packed range that comes from wiretype), but ultimately I've disabled it as it is more complicated in some cases. I think I will remove this comment for now and left only TODO, as it is not valid for all cases.

File: systemverilog-plugin/UhdmAst.cc [hunk 4]
```diff
@@ -962,29 +962,47 @@ static AST::AstNode *convert_dot(AST::AstNode *wire_node, AST::AstNode *node, AS
...
-            expanded->children[0] = new AST::AstNode(
-              AST::AST_ADD, expanded->children[0],
-              new AST::AstNode(AST::AST_MUL, AST::AstNode::mkconst_int(struct_size_int, true, 32), node->children[0]->children[0]->clone()));
+    std::vector<AST::AstNode *> struct_ranges;
+    for (auto c : node->children) {
+        if (c->type == AST::AST_RANGE) {
+            struct_ranges.push_back(c);
         }
     }
+    const auto wire_node_unpacked_ranges_size = wire_node->attributes[UhdmAst::unpacked_ranges()]->children.size();
```
[the remark this answers is not in the document]
Comment (kamilrakoczy) [reply systemverilog-plugin/UhdmAst.cc:974]: At this point, ``wire_node`` should have it, I've added assert to make sure it is always present.

[the remark this answers is not in the document]
Comment (wsipak) [reply systemverilog-plugin/UhdmAst.cc:974]: Thanks!

**Review by wsipak — APPROVED** [review]

File: systemverilog-plugin/UhdmAst.cc [hunk 5]
```diff
@@ -962,29 +962,49 @@ static AST::AstNode *convert_dot(AST::AstNode *wire_node, AST::AstNode *node, AS
...
+    // log_assert(struct_ranges.size() == (wire_node_ranges_size - 1));
+
+    // Get size of single structure
+    int struct_size_int = get_max_offset_struct(struct_node) + 1;
+    // Get reverse iterator of wire dimensions
+    auto dimensions_it = wire_node->multirange_dimensions.rbegin();
+    unsigned long range_id = 0;
+    for (auto it = struct_ranges.rbegin(); it != struct_ranges.rend(); it++) {
+        // in 'dot' context, we need to select specific struct
+        // element, so assert that there is only 1 child (range with single const) in struct
```
Comment (mandrys) [inline systemverilog-plugin/UhdmAst.cc:990]: Here we assert that an item in `struct_ranges` has only 1 child, but the comment says that we expect only 1 child in a struct - is that right or I'm messing sth up? Generally this comment is not clear for me - could you explain, why this assert is important for the fact that we select a specific struct in 'dot' context?

Comment (kamilrakoczy) [reply systemverilog-plugin/UhdmAst.cc]: Lets consider such assign:
```systemverilog
assign group_status[0][0].a = 1'b1;
```
At this point, ``.a`` part was already expanded by ``expand_dot`` function, and here we are reverse-iterating ranges of ``group_status``.

This assert checks, that each range have only 1 child, so it is range with single const, not part-select (e.g. ``[1:0]``) as latter is not valid systemverilog code.

I slightly changed this comment, I hope it is more clear now.

Comment (mandrys) [reply systemverilog-plugin/UhdmAst.cc]: OK, thanks, now I see what you meant.

**Review by mandrys — APPROVED** [review]

File: systemverilog-plugin/UhdmAst.cc [hunk 6]
```diff
@@ -962,29 +962,49 @@ static AST::AstNode *convert_dot(AST::AstNode *wire_node, AST::AstNode *node, AS
...
+                      ? new AST::AstNode(AST::AST_SUB, AST::AstNode::mkconst_int(*dimensions_it - 1, true, 32), (*it)->children[0]->clone())
+                      : (*it)->children[0]->clone();
+        // calculate which struct we selected
+        auto move_offset = new AST::AstNode(AST::AST_MUL, AST::AstNode::mkconst_int(struct_size_int, true, 32), elem);
+        // move our expanded dot to currently selected struct
+        expanded->children[0] = new AST::AstNode(AST::AST_ADD, move_offset->clone(), expanded->children[0]);
+        expanded->children[1] = new AST::AstNode(AST::AST_ADD, move_offset, expanded->children[1]);
+        // next range, multiply struct_size_int by whole dimension size
+        struct_size_int *= *dimensions_it;
+        // move iterator to next dimension width
```
Comment (mglb) [inline systemverilog-plugin/UhdmAst.cc:1004]: Obvious comment. You can optionally mention why you move iterator by 2, like "multirange_dimensions stores interleaved offset and size. Move to next dimension's size"

File: systemverilog-plugin/UhdmAst.cc [hunk 7]
```diff
@@ -962,29 +962,49 @@ static AST::AstNode *convert_dot(AST::AstNode *wire_node, AST::AstNode *node, AS
...
+    // but wiretype ranges contains also struct range that is already expanded in 'expand_dot'
+    // we need to find a way to calculate size of wire ranges without struct range here to enable this assert
+    // const auto wire_node_packed_ranges_size = wire_node->attributes[UhdmAst::packed_ranges()]->children.size();
+    // const auto wire_node_ranges_size = wire_node_packed_ranges_size + wire_node_unpacked_ranges_size;
+    // log_assert(struct_ranges.size() == (wire_node_ranges_size - 1));
+
+    // Get size of single structure
+    int struct_size_int = get_max_offset_struct(struct_node) + 1;
+    // Get reverse iterator of wire dimensions
+    auto dimensions_it = wire_node->multirange_dimensions.rbegin();
```
[the remark this answers is not in the document]
Comment (mglb) [reply systemverilog-plugin/UhdmAst.cc:986]: Also, please rename it to something like `wire_dimension_size_it`,

File: systemverilog-plugin/UhdmAst.cc [hunk 8]
```diff
@@ -962,29 +962,49 @@ static AST::AstNode *convert_dot(AST::AstNode *wire_node, AST::AstNode *node, AS
...
+        // if unpacked range, select from back;
+        auto elem = is_unpacked_range
+                      ? new AST::AstNode(AST::AST_SUB, AST::AstNode::mkconst_int(*dimensions_it - 1, true, 32), (*it)->children[0]->clone())
+                      : (*it)->children[0]->clone();
+        // calculate which struct we selected
+        auto move_offset = new AST::AstNode(AST::AST_MUL, AST::AstNode::mkconst_int(struct_size_int, true, 32), elem);
+        // move our expanded dot to currently selected struct
+        expanded->children[0] = new AST::AstNode(AST::AST_ADD, move_offset->clone(), expanded->children[0]);
+        expanded->children[1] = new AST::AstNode(AST::AST_ADD, move_offset, expanded->children[1]);
+        // next range, multiply struct_size_int by whole dimension size
```
Comment (mglb) [inline systemverilog-plugin/UhdmAst.cc:1002]: Redundant comment.

**Review by mglb — COMMENTED** [review]
Logic looks good, but this was quite hard to follow. What (I think) would make it more readable:
- brief description how the input (`expand_dot` output) looks, and what has to be done. Really short, kind of "We have something like this, and have to extract this and that and glue it together into a tree like this".
- remove comments with tautology
- more precise names, which would make few comments redundant.

**Pull request merged**